### PR TITLE
Add ClassCodec with parameterized field support

### PR DESCRIPTION
**Pull request by pnwpedro**
### Description

* Remove MappingContext from the codec API (🎉), soon to be removed from the library
* Move MappingInfo directly into the ClassCodec. These codecs are cached, so it feels like it simplifies things for now.
* Advance the underlying JSON token stream at Parser instantiation. This leaves the None token a concern of the parser, which can eventually be used to determine whether or not we're consuming a partial or full token stream.
* Add BaseCodec, ClassCodec, IntCodec, StringCodec, ListCodec, MapCodec. These codecs allowed me to implement support parameterized lists and maps as fields in the ClassCodec!
* Add typeArgument to the FaunaField annotation. I chose a single type argument rather than a list because (at least for now), we will not support parameterized types with more than one type arg.
* Implements a DefaultCodecProvider and DefaultCodecRegistry. These APIs changed a little bit and I'm not sure I'm perfectly happy with them yet, but don't want to waffle too much on these. They work for now.
* Implements DynamicCodec. This is lightly tested outside of some E2E tests I was doing locally, but I didn't want to pull it out. Tests to follow.

##### Class with Generic Fields

We use the new typeArgument property of the FaunaField annotation to tell the codec how to handle the elements of a list and values of a map.
```
@FaunaObject
public class Foo {

    @FaunaField(name = "bar")
    public String bar = "baz";

    @FaunaField(name = "a_list", typeArgument = String.class)
    public List<String> list  = List.of("item1");

    @FaunaField(name = "a_map", typeArgument = Integer.class)
    public Map<String,Integer> map  = Map.of("key1", 42);
}
```

### Motivation and context

Proves out a major part of the refactor, which was to implement straight forward support for encoding classes with list/map fields.

This sets the stage for implementation of the remaining codecs.

### How was the change tested?

* Added light unit tests
* Locally I piped the codec API through the client and tested a few of the cases E2E.

### Change types

* - [ ] Bug fix (non-breaking change that fixes an issue)
* - [X] New feature (non-breaking change that adds functionality)
* - [ ] Breaking change (backwards-incompatible fix or feature)

### Checklist:

* - [X] My code follows the code style of this project.
* - [ ] My change requires a change to Fauna documentation.
* - [ ] My change requires a change to the README, and I have updated it accordingly.

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/fauna/codec/codecs/ClassCodec.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        String id = null;
+        String name = null;
+        Module coll = null;
+        boolean exists = true;
+        String cause = null;
+
+        while (parser.read() && parser.getCurrentTokenType() != endToken) {
+            if (parser.getCurrentTokenType() != FaunaTokenType.FIELD_NAME) {
+                throw new ClientException(unexpectedTokenExceptionMessage(parser.getCurrentTokenType()));
+            }
```
Comment (findgriffin) [inline src/main/java/com/fauna/codec/codecs/ClassCodec.java:138]: This method is a little hard to follow, and having this many nulls at the beginning gives me the heebie jeebies, but I'm sure it works :)

File: src/main/java/com/fauna/codec/codecs/DynamicCodec.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        this.provider = provider;
+    }
+
+    @Override
+    public Object decode(UTF8FaunaParser parser) throws IOException {
+        Object value = null;
+        switch (parser.getCurrentTokenType()) {
+            case NULL:
+                break;
+            case START_OBJECT:
```
Comment (findgriffin) [inline src/main/java/com/fauna/codec/codecs/DynamicCodec.java:35]: A strategy I've sometimes used because Java is dumb and makes you have the break statement is to return directly from the switch statement. It seems cleaner to me, but again, just an opinion.

File: src/main/java/com/fauna/codec/codecs/PageCodec.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    @Override
+    public void encode(UTF8FaunaGenerator gen, L obj) throws IOException {
+        if (obj == null) {
+            gen.writeNullValue();
+            return;
+        }
+
+        throw new ClientException(this.unsupportedTypeMessage(obj.getClass()));
```
Comment (findgriffin) [inline src/main/java/com/fauna/codec/codecs/PageCodec.java:44]:
```suggestion
        } else {
            throw new ClientException(this.unsupportedTypeMessage(obj.getClass()));
        }
```

**Review by findgriffin — APPROVED** [review]
Awesome!

File: src/main/java/com/fauna/codec/DefaultCodecRegistry.java [hunk 4]
```diff
@@ -1,13 +1,11 @@
 package com.fauna.codec;
 
-import java.util.HashMap;
+import java.util.concurrent.ConcurrentHashMap;
 
 public class DefaultCodecRegistry implements CodecRegistry {
+    private static final ConcurrentHashMap<CodecRegistryKey, Codec<?>> codecs = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (pnwpedro) [reply src/main/java/com/fauna/codec/DefaultCodecRegistry.java:6]: These are reasonable considerations.

1. Yes, we should think of them as immutable
2. This isn't quite true. We add class codec instances to the cache as we build them. I'm not sure we can make assumptions about concurrency, but simply need this to be thread-safe.
3-4. Based on a quick search, it looks like read is pretty similar. Write takes the big hit as expected. https://stackoverflow.com/questions/1378310/performance-concurrenthashmap-vs-hashmap

That said I realize that I made a mistake. I'm calling synchronized in the provider, and backing the registry with a concurrent hash map–it's mixed mode. I'll stick with the concurrent hashmap and removed the synchronized behavior. Obv let me know if you have strong feels here and we can discuss a bit.

File: src/main/java/com/fauna/codec/DefaultCodecProvider.java [hunk 5]
```diff
@@ -1,25 +1,60 @@
...
+                }
+            }
         }
 
         return registry.get(key);
     }
 
-    private <T,K> Codec<T> generate(Class<T> clazz, Class<K> subClazz) {
-        return null;
+    @SuppressWarnings({"unchecked"})
```
[the remark this answers is not in the document]
Comment (pnwpedro) [reply src/main/java/com/fauna/codec/DefaultCodecProvider.java:43]: In this case there were several, but I'll look to narrow in other places.

File: src/main/java/com/fauna/codec/codecs/ClassCodec.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        return null;
+    }
+
+    private Object createInstance() {
+        try {
+            Class<?> clazz = Class.forName(type.getTypeName());
+            Constructor<?> constructor = clazz.getConstructor();
+            return constructor.newInstance();
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
```
[the remark this answers is not in the document]
Comment (pnwpedro) [reply src/main/java/com/fauna/codec/codecs/ClassCodec.java:123]: yeah, i agree. will bump it up a level

File: src/main/java/com/fauna/codec/codecs/ClassCodec.java (hunk 1 above)
Comment (pnwpedro) [reply src/main/java/com/fauna/codec/codecs/ClassCodec.java:138]: I def agree, although this is a port from the existing. I'll take a quick look to see if I can come up with a straight forward rewrite

File: src/main/java/com/fauna/codec/DefaultCodecRegistry.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (freels) [reply src/main/java/com/fauna/codec/DefaultCodecRegistry.java:6]: Alternatively, you could make the cache an unmodifiable map stored in a volatile, and copy it on update. (put would have to be synchronized).

[the remark this answers is not in the document]
Comment (pnwpedro) [reply src/main/java/com/fauna/codec/DefaultCodecRegistry.java:6]: > Alternatively, you could make the cache an unmodifiable map stored in a volatile, and copy it on update. (put would still have to be synchronized).

Cool, good to have some options in the back pocket, but maybe we can optimize this when it needs optimizing? However if there's a strong affinity to a particular solution now, I'll follow suite. Just let me know. @freels / @findgriffin

[the remark this answers is not in the document]
Comment (freels) [reply src/main/java/com/fauna/codec/DefaultCodecRegistry.java:6]: IME going through a volatile (for reads) is going to scale to more threads than going through synchronized (the JVM is very good at optimizing away the lock in single-threaded cases, tho). CHM fits the bill here.

File: src/main/java/com/fauna/codec/CodecRegistryKey.java [hunk 7]
```diff
@@ -1,16 +1,38 @@
...
     }
 
-    // TODO: Override equals/hash
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        } else if (other instanceof CodecRegistryKey) {
+            CodecRegistryKey otherCRK = (CodecRegistryKey) other;
+            return this.base == otherCRK.base && this.typeArg == otherCRK.typeArg;
```
Comment (freels) [inline src/main/java/com/fauna/codec/CodecRegistryKey.java:28]: I _think_ you want to call equals here

```suggestion
            return this.base.equals(otherCRK.base) && this.typeArg.equals(otherCRK.typeArg);
```

**Review by freels — COMMENTED** [review]
overall looks good. One comment about equality impl

File: src/main/java/com/fauna/codec/CodecRegistryKey.java (hunk 7 above)
Comment (pnwpedro) [reply src/main/java/com/fauna/codec/CodecRegistryKey.java]: Good catch. I did something slightly different because the typeArg can be null.

**Review by findgriffin — APPROVED** [review]
